test/usecases/table/simulate.py

```diff
@@ -0,0 +1,170 @@
+from neuron import gui
+from neuron import h
+
+
+def setup_sim():
+    section = h.Section()
+    section.insert("tbl")
+
+    return section
+
+
+def check_table(y_no_table, y_table, rtol1=1e-2, rtol2=1e-8):
+    assert np.allclose(y_no_table, y_table, rtol=rtol1), f"{y_no_table} != {y_table}"
+    # if the table is not used, we should get identical results, but we don't,
+    # hence the assert below
+    assert not np.allclose(
+        y_no_table, y_table, rtol=rtol2
+    ), f"{y_no_table} != {y_table}"
+
+
+def test_function():
+    section = setup_sim()
+
+    x = np.linspace(-3, 5, 500)
+
+    func = section(0.5).tbl.example_function
+
+    h.c1_tbl = 1
+    h.c2_tbl = 2
+
+    h.usetable_tbl = 0
+    y_no_table = np.array([func(i) for i in x])
+
+    h.usetable_tbl = 1
+    y_table = np.array([func(i) for i in x])
+
+    check_table(y_table, y_no_table, rtol1=1e-4)
+
+    # verify that the table just "clips" the values outside of the range
+    assert func(x[0] - 10) == y_table[0]
+    assert func(x[-1] + 10) == y_table[-1]
+
+    # change parameters and verify
+    h.c1_tbl = 3
+    h.c2_tbl = 4
+
+    h.usetable_tbl = 0
+    y_params_no_table = np.array([func(i) for i in x])
+
+    h.usetable_tbl = 1
+    y_params_table = np.array([func(i) for i in x])
+
+    check_table(y_params_table, y_params_no_table, rtol1=1e-4)
+
+
+def test_procedure():
+    section = setup_sim()
+
+    x = np.linspace(-4, 6, 300)
+
+    proc = section(0.5).tbl.example_procedure
+
+    def modified_args(arg):
+        """
+        Pass arg to the procedure and return new values
+        """
+        proc(arg)
+        return section(0.5).tbl.v1, section(0.5).tbl.v2
+
+    h.c1_tbl = 1
+    h.c2_tbl = 2
+
+    h.usetable_tbl = 0
+
+    v1_no_table, v2_no_table = np.transpose(np.array([modified_args(i) for i in x]))
+
+    h.usetable_tbl = 1
+    v1_table, v2_table = np.transpose(np.array([modified_args(i) for i in x]))
+
+    check_table(v1_table, v1_no_table, rtol1=1e-3)
+    check_table(v2_table, v2_no_table, rtol1=1e-3)
+
+    # verify that the table just "clips" the values outside of the range
+    v1, v2 = modified_args(x[0] - 10)
+    assert v1 == v1_table[0] and v2 == v2_table[0]
+    v1, v2 = modified_args(x[-1] + 10)
+    assert v1 == v1_table[-1] and v2 == v2_table[-1]
+
+    # change params and verify
+    # N.B. since this controls the frequency of the oscillations (as our
+    # procedure is a trig function), a higher value of the frequency degrades
+    # the accuracy of the interpolation
+    h.c1_tbl = 0.1
+    h.c2_tbl = 0.3
+
+    h.usetable_tbl = 0
+    v1_params_no_table, v2_params_no_table = np.transpose(
+        np.array([modified_args(i) for i in x])
+    )
+
+    h.usetable_tbl = 1
+    v1_params_table, v2_params_table = np.transpose(
+        np.array([modified_args(i) for i in x])
+    )
+
+    check_table(v1_params_table, v1_params_no_table, rtol1=1e-3)
+    check_table(v2_params_table, v2_params_no_table, rtol1=1e-3)
```

Comment:
If you feel like playing golf, the preceeding 15 lines, are repeated from above. You could rework as:
```
   check_proc_table(1, 2)
   check_proc_table(0.1, 0.3)
```
Then you notice that if you then keep `v1` and `v2` in a single array, you can play golf again; and probably get it to also work for functions too. The only thing that differs is how you call the function/procedure. Hence you could inject that as a callback:
```
def compare_table_and_no_table(c1, c2, modified_args):
    h.c1_tbl = c1
    h.c2_tbl = c2

    h.usetable_tbl = 0
    values_no_table = np.array([modified_args(i) for i in x])

    h.usetable_tbl = 1
    values_with_table =(
        np.array([modified_args(i) for i in x])
    )

    check_table(values_with_table, values_no_table, rtol1=1e-3)
```
Then one round of renaming:
* `compare_table_and_no_table` would be `check_table`.
* `check_table` would unfortunately have to turn into `is_{similar, close}`. You might be able to inline it.

This way you can collapse four copy-paste repetitions into a single function, called four times.